### PR TITLE
James 2262: Add more testing for SpamAssassin and documentation how to use SpamAssassin

### PR DESCRIPTION
**Pull request by quynhn**
The document is there on SpamAssassin mailet

### CODE REVIEW

File: server/mailet/integration-testing/src/test/java/org/apache/james/transport/mailets/SpamAssassinTest.java [hunk 1]
```diff
@@ -0,0 +1,213 @@
...
+    private static final int IMAP_PORT = 1143;
+    private static final int SMTP_PORT = 1025;
+
+    public static final String JAMES_ORG = "james.org";
+    public static final String SENDER = "sender@" + JAMES_ORG;
+    public static final String TO = "to@" + JAMES_ORG;
+    public static final String PASSWORD = "secret";
+
+    public SwarmGenericContainer spamAssassinContainer = new SwarmGenericContainer(ContainerNames.SPAMASSASSIN)
+        .portBinding(783, 783)
```
Comment (chibenwa) [inline server/mailet/integration-testing/src/test/java/org/apache/james/transport/mailets/SpamAssassinTest.java]: Can we avoid this port binding? (See below)

Comment (chibenwa) [reply server/mailet/integration-testing/src/test/java/org/apache/james/transport/mailets/SpamAssassinTest.java]: We miss a wait strategy...

Like:

```
        .withExposedPorts(783)
        .waitingFor(new HostPortWaitStrategy())
```

File: server/mailet/integration-testing/src/test/java/org/apache/james/transport/mailets/SpamAssassinTest.java [hunk 2]
```diff
@@ -0,0 +1,213 @@
...
+                        .matcher(RecipientIsLocal.class)
+                        .mailet(VacationMailet.class)
+                        .build())
+                    .addMailet(MailetConfiguration.builder()
+                        .matcher(RecipientIsLocal.class)
+                        .mailet(LocalDelivery.class)
+                        .build())
+                    .build())
+            .build();
+        jamesServer = TemporaryJamesServer.builder().build(temporaryFolder, mailets);
```
Comment (chibenwa) [inline server/mailet/integration-testing/src/test/java/org/apache/james/transport/mailets/SpamAssassinTest.java]: Use base module MemoryJamesServer.SMTP_AND_IMAP

See GatewayRemoteDeliveryIntegrationTest

File: server/mailet/integration-testing/src/test/java/org/apache/james/transport/mailets/SpamAssassinTest.java [hunk 3]
```diff
@@ -0,0 +1,213 @@
...
+                        .addProperty("simplemapping",
+                            SpamAssassinInvoker.FLAG_MAIL_ATTRIBUTE_NAME + ";" + SpamAssassinInvoker.FLAG_MAIL_ATTRIBUTE_NAME + "," +
+                            SpamAssassinInvoker.STATUS_MAIL_ATTRIBUTE_NAME + ";" + SpamAssassinInvoker.STATUS_MAIL_ATTRIBUTE_NAME)
+                        .build())
+                    .addMailet(MailetConfiguration.builder()
+                        .matcher(All.class)
+                        .mailet(RemoveMimeHeader.class)
+                        .addProperty("name", "bcc")
+                        .build())
+                    .addMailet(MailetConfiguration.builder()
```
Comment (chibenwa) [inline server/mailet/integration-testing/src/test/java/org/apache/james/transport/mailets/SpamAssassinTest.java]: If you us the right baseModule, this mailet is not required...


See GatewayRemoteDeliveryIntegrationTest

File: server/mailet/integration-testing/src/test/java/org/apache/james/transport/mailets/SpamAssassinTest.java [hunk 4]
```diff
@@ -0,0 +1,213 @@
...
+        }
+    }
+
+    @Test
+    public void spamAssassinShouldAppendNewHeaderWhichDetectIsSpamWhenSpamMessage() throws Exception {
+        MimeMessage message = MimeMessageBuilder.mimeMessageBuilder()
+            .setSender(SENDER)
+            .addToRecipient(TO)
+            .setSubject("This is the subject")
+            .setText("XJS*C4JDBQADN1.NSBN3*2IDNEN*GTUBE-STANDARD-ANTI-UBE-TEST-EMAIL*C.34X")
```
Comment (chibenwa) [inline server/mailet/integration-testing/src/test/java/org/apache/james/transport/mailets/SpamAssassinTest.java]: Maybe extract it in a variable to explain it is Spam...

File: server/mailet/mailets/src/main/java/org/apache/james/transport/mailets/SpamAssassin.java [hunk 5]
```diff
@@ -65,18 +72,18 @@
      * @see org.apache.mailet.base.GenericMailet#init()
      */
     public void init() throws MessagingException {
-        spamdHost = getInitParameter("spamdHost");
-        if (spamdHost == null || spamdHost.equals("")) {
-            spamdHost = "127.0.0.1";
+        spamdHost = getInitParameter(SPAMD_HOST);
+        if (StringUtils.isEmpty(spamdHost)) {
+            spamdHost = DEFAULT_HOST;
```
Comment (chibenwa) [inline server/mailet/mailets/src/main/java/org/apache/james/transport/mailets/SpamAssassin.java]: String re-allocation spotted ;-)

Maybe we can use an optional here:

```
spamHost = Optional.ofNullable(getInitParameter(SPAMD_HOST))
    .filter(s -> !Strings.isNullOrEmpty(s))
    .orElse(DEFAULT_HOST);
```

File: server/mailet/mailets/src/main/java/org/apache/james/transport/mailets/SpamAssassin.java [hunk 6]
```diff
@@ -65,18 +72,18 @@
...
         }
 
-        String port = getInitParameter("spamdPort");
-        if (port == null || port.equals("")) {
-            spamdPort = 783;
+        String port = getInitParameter(SPAMD_PORT);
```
Comment (chibenwa) [inline server/mailet/mailets/src/main/java/org/apache/james/transport/mailets/SpamAssassin.java]:
```
spamPort = MailetUtil.getInitParameterAsStrictlyPositiveInteger(getInitParameter(SPAMD_PORT), DEFAULT_PORT);
```

Comment (chibenwa) [reply server/mailet/mailets/src/main/java/org/apache/james/transport/mailets/SpamAssassin.java]: could we add a check to throw a `MessageException` when the port is higher that `65535`?

File: server/mailet/mailets/src/test/java/org/apache/james/transport/mailets/SpamAssassinTest.java [hunk 7]
```diff
@@ -0,0 +1,179 @@
...
+            .setProperty(SpamAssassin.SPAMD_PORT, "" + spamPort)
+            .build());
+
+        assertThat(mailet.getSpamdPort()).isEqualTo(spamPort);
+    }
+
+    @Test
+    public void serviceShouldWriteSpamAttributeOnMail() throws Exception {
+        int port = PortUtil.getNonPrivilegedPort();
+        new Thread(new MockSpamd(port)).start();
```
Comment (chibenwa) [inline server/mailet/mailets/src/test/java/org/apache/james/transport/mailets/SpamAssassinTest.java]: If you start it, then we need to stop it at the end of your test. What about starting it / stopping it using @Before @After ?

And wrap your thread in an Executor (fixed threadpool something?)

Comment (chibenwa) [reply server/mailet/mailets/src/test/java/org/apache/james/transport/mailets/SpamAssassinTest.java]: Pro tip:
 - Pass port 0 to get a random port
 - Add a getPort method to MockSpamD

File: server/mailet/mailets/src/test/java/org/apache/james/transport/mailets/SpamAssassinTest.java [hunk 8]
```diff
@@ -0,0 +1,179 @@
...
+            .setProperty(SpamAssassin.SPAMD_PORT, "noNumber")
+            .build())).isInstanceOf(MessagingException.class);
+    }
+
+    @Test
+    public void initShouldSetSpamPortWhenPresent() throws Exception {
+        int spamPort = 1000;
+        mailet.init(FakeMailetConfig.builder()
+            .mailetName("SpamAssassin")
+            .setProperty(SpamAssassin.SPAMD_PORT, "" + spamPort)
```
Comment (chibenwa) [inline server/mailet/mailets/src/test/java/org/apache/james/transport/mailets/SpamAssassinTest.java]: String.valueOf(spamPort)

File: server/mailet/mailets/src/test/java/org/apache/james/transport/mailets/SpamAssassinTest.java [hunk 9]
```diff
@@ -0,0 +1,179 @@
...
+
+    @Test
+    public void serviceShouldWriteMessageAsNotSpamWhenNotSpam() throws Exception {
+        int port = PortUtil.getNonPrivilegedPort();
+        new Thread(new MockSpamd(port)).start();
+
+        FakeMailetConfig mailetConfiguration = FakeMailetConfig.builder()
+            .mailetName("SpamAssassin")
+            .setProperty(SpamAssassin.SPAMD_HOST, "localhost")
+            .setProperty(SpamAssassin.SPAMD_PORT, "" + port)
```
Comment (chibenwa) [inline server/mailet/mailets/src/test/java/org/apache/james/transport/mailets/SpamAssassinTest.java]: idem

File: server/mailet/mailets/src/test/java/org/apache/james/transport/mailets/SpamAssassinTest.java [hunk 10]
```diff
@@ -0,0 +1,179 @@
...
+
+    @Test
+    public void serviceShouldWriteMessageAsSpamWhenSpam() throws Exception {
+        int port = PortUtil.getNonPrivilegedPort();
+        new Thread(new MockSpamd(port)).start();
+
+        FakeMailetConfig mailetConfiguration = FakeMailetConfig.builder()
+            .mailetName("SpamAssassin")
+            .setProperty(SpamAssassin.SPAMD_HOST, "localhost")
+            .setProperty(SpamAssassin.SPAMD_PORT, "" + port)
```
Comment (chibenwa) [inline server/mailet/mailets/src/test/java/org/apache/james/transport/mailets/SpamAssassinTest.java]: idem

File: server/mailet/mailets/src/test/java/org/apache/james/transport/mailets/SpamAssassinTest.java [hunk 11]
```diff
@@ -0,0 +1,179 @@
...
+import org.apache.mailet.base.test.FakeMail;
+import org.apache.mailet.base.test.FakeMailetConfig;
+import org.apache.mailet.base.test.MimeMessageBuilder;
+import org.junit.Test;
+
+public class SpamAssassinTest {
+    private SpamAssassin mailet = new SpamAssassin();
+
+    @Test
+    public void initShouldNotThrowWhenNoSpamHost() throws Exception {
```
Comment (chibenwa) [inline server/mailet/mailets/src/test/java/org/apache/james/transport/mailets/SpamAssassinTest.java]: Here it is throwing...

This test don't like useful, can we remove it?

File: server/mailet/mailets/src/test/java/org/apache/james/transport/mailets/SpamAssassinTest.java [hunk 12]
```diff
@@ -0,0 +1,179 @@
...
+    public void initShouldSetSpamdHostWhenPresent() throws Exception {
+        String spamdHost = "any.host";
+        mailet.init(FakeMailetConfig.builder()
+            .mailetName("SpamAssassin")
+            .setProperty(SpamAssassin.SPAMD_HOST, spamdHost)
+            .build());
+
+        assertThat(mailet.getSpamdHost()).isEqualTo(spamdHost);
+    }
+
```
Comment (chibenwa) [inline server/mailet/mailets/src/test/java/org/apache/james/transport/mailets/SpamAssassinTest.java]: We need also a test for

```
+        mailet.init(FakeMailetConfig.builder()
+            .mailetName("SpamAssassin", "")
+            .build());
```

File: server/mailet/mailets/src/test/java/org/apache/james/transport/mailets/SpamAssassinTest.java [hunk 13]
```diff
@@ -0,0 +1,179 @@
...
+    @Test
+    public void initShouldThrowWhenSpamdPortIsNotNumber() throws Exception {
+        assertThatThrownBy(() -> mailet.init(FakeMailetConfig.builder()
+            .mailetName("SpamAssassin")
+            .setProperty(SpamAssassin.SPAMD_PORT, "noNumber")
+            .build())).isInstanceOf(MessagingException.class);
+    }
+
```
Comment (chibenwa) [inline server/mailet/mailets/src/test/java/org/apache/james/transport/mailets/SpamAssassinTest.java]: We need a test as well for 

```
+        assertThatThrownBy(() -> mailet.init(FakeMailetConfig.builder()
+            .mailetName("SpamAssassin")
+            .setProperty(SpamAssassin.SPAMD_PORT, "")
+            .build())).isInstanceOf(MessagingException.class);
```

Comment (chibenwa) [reply server/mailet/mailets/src/test/java/org/apache/james/transport/mailets/SpamAssassinTest.java]: What happens with

```
+        assertThatThrownBy(() -> mailet.init(FakeMailetConfig.builder()
+            .mailetName("SpamAssassin")
+            .setProperty(SpamAssassin.SPAMD_PORT, "0")
+            .build())).isInstanceOf(MessagingException.class);
```

Or

```
+        assertThatThrownBy(() -> mailet.init(FakeMailetConfig.builder()
+            .mailetName("SpamAssassin")
+            .setProperty(SpamAssassin.SPAMD_PORT, "-1")
+            .build())).isInstanceOf(MessagingException.class);
```

Or

```
+        assertThatThrownBy(() -> mailet.init(FakeMailetConfig.builder()
+            .mailetName("SpamAssassin")
+            .setProperty(SpamAssassin.SPAMD_PORT, "1000000000")
+            .build())).isInstanceOf(MessagingException.class);
```

File: server/mailet/mailets/src/main/java/org/apache/james/transport/mailets/SpamAssassin.java [hunk 14]
```diff
@@ -57,30 +64,24 @@
 @Experimental
```
Comment (aduprat) [inline server/mailet/mailets/src/main/java/org/apache/james/transport/mailets/SpamAssassin.java]: Is it still experimental?

File: server/mailet/integration-testing/src/test/java/org/apache/james/transport/mailets/SpamAssassinTest.java [hunk 15]
```diff
@@ -0,0 +1,215 @@
...
+
+    public SwarmGenericContainer spamAssassinContainer = new SwarmGenericContainer(ContainerNames.SPAMASSASSIN)
+        .withExposedPorts(783)
+        .withAffinityToContainer()
+        .waitingFor(new HostPortWaitStrategy());
+
+    public TemporaryFolder temporaryFolder = new TemporaryFolder();
+
+    @Rule
+    public final RuleChain chain = RuleChain.outerRule(temporaryFolder).around(spamAssassinContainer);
```
[the remark this answers is not in the document]
Comment (chibenwa) [reply server/mailet/integration-testing/src/test/java/org/apache/james/transport/mailets/SpamAssassinTest.java]: Good catch

File: server/mailet/mailets/src/test/java/org/apache/james/transport/mailets/SpamAssassinTest.java (hunk 7 above)
Comment (strokyl) [reply server/mailet/mailets/src/test/java/org/apache/james/transport/mailets/SpamAssassinTest.java]: Would be even better to write a JUNIT rule for that no ?

Comment (chibenwa) [reply server/mailet/mailets/src/test/java/org/apache/james/transport/mailets/SpamAssassinTest.java]: https://github.com/linagora/james-project/pull/1194/commits/ac79e58b211c7ad8641a72c02d98e8b561afda33 

```
JAMES-2262 Use a Test rule for Mock SpamD test rule 
```

Come on!

File: server/mailet/integration-testing/src/test/java/org/apache/james/transport/mailets/SpamAssassinTest.java [hunk 16]
```diff
@@ -133,35 +127,89 @@ public void tearDown() {
...
+
+            String receivedHeaders = imapMessageReader.readFirstMessageInInbox(TO, PASSWORD);
+
+            assertThat(receivedHeaders).contains(SpamAssassinInvoker.FLAG_MAIL_ATTRIBUTE_NAME + ": YES");
+            assertThat(receivedHeaders).contains(SpamAssassinInvoker.STATUS_MAIL_ATTRIBUTE_NAME + ": Yes");
         }
     }
 
+    @Test
+    public void spamAssassinShouldAppendNewHeaderWhichNoWhenNonSpamMessage() throws Exception {
```
Comment (strokyl) [inline server/mailet/integration-testing/src/test/java/org/apache/james/transport/mailets/SpamAssassinTest.java]: s/WhichNo/WithNoValue/

File: server/mailet/integration-testing/src/test/java/org/apache/james/transport/mailets/SpamAssassinTest.java [hunk 17]
```diff
@@ -133,35 +127,89 @@ public void tearDown() {
...
+             IMAPMessageReader imapMessageReader = new IMAPMessageReader(LOCALHOST_IP, IMAP_PORT)) {
+
+            messageSender.sendMessage(mail);
+
+            calmlyAwait.atMost(Duration.ONE_MINUTE).until(messageSender::messageHasBeenSent);
+            calmlyAwait.atMost(Duration.ONE_MINUTE).until(() -> imapMessageReader.userReceivedMessage(TO, PASSWORD));
+
+            String receivedHeaders = imapMessageReader.readFirstMessageInInbox(TO, PASSWORD);
+
+            assertThat(receivedHeaders).contains(SpamAssassinInvoker.FLAG_MAIL_ATTRIBUTE_NAME + ": NO");
```
Comment (strokyl) [inline server/mailet/integration-testing/src/test/java/org/apache/james/transport/mailets/SpamAssassinTest.java]: NO or No ?

File: server/container/util-java8/src/main/java/org/apache/james/util/Port.java [hunk 18]
```diff
@@ -0,0 +1,43 @@
...
+    public static final int MAX_PORT_VALUE = 65535;
+    public static final int PRIVILEGED_PORT_BOUND = 1024;
+    private static final Range<Integer> VALID_PORT_RANGE = Range.closed(1, MAX_PORT_VALUE);
+
+    public static int generateValidUnprivilegedPort() {
+        return new Random().nextInt(Port.MAX_PORT_VALUE - PRIVILEGED_PORT_BOUND) + PRIVILEGED_PORT_BOUND;
+    }
+
+    public static void assertValid(int port) {
+        Preconditions.checkArgument(isValid(port), "Port should be between 0 and 65535");
```
Comment (rouazana) [inline server/container/util-java8/src/main/java/org/apache/james/util/Port.java]: isValid does not accept port = 0

File: server/mailet/integration-testing/src/test/java/org/apache/james/transport/mailets/SpamAssassinTest.java (hunk 16 above)
Comment (chibenwa) [reply server/mailet/integration-testing/src/test/java/org/apache/james/transport/mailets/SpamAssassinTest.java]: Outdated

File: server/mailet/integration-testing/src/test/java/org/apache/james/transport/mailets/SpamAssassinTest.java (hunk 17 above)
Comment (chibenwa) [reply server/mailet/integration-testing/src/test/java/org/apache/james/transport/mailets/SpamAssassinTest.java]: NO.

Sorry, but I don't intend to change this, as I don't intend to break retro-compatibility without a good reason.

File: server/mailet/integration-testing/src/test/java/org/apache/james/transport/mailets/SpamAssassinTest.java [hunk 19]
```diff
@@ -0,0 +1,215 @@
...
+
+            String receivedHeaders = imapMessageReader.readFirstMessageInInbox(TO, PASSWORD);
+
+            assertThat(receivedHeaders).contains(SpamAssassinInvoker.FLAG_MAIL_ATTRIBUTE_NAME + ": YES");
+            assertThat(receivedHeaders).contains(SpamAssassinInvoker.STATUS_MAIL_ATTRIBUTE_NAME + ": Yes");
+        }
+    }
+
+    @Test
+    public void spamAssassinShouldAppendNewHeaderWhichNoWhenNonSpamMessage() throws Exception {
```
Comment (rouazana) [inline server/mailet/integration-testing/src/test/java/org/apache/james/transport/mailets/SpamAssassinTest.java]: s/Which/With

File: server/container/util-java8/src/test/java/org/apache/james/util/PortTest.java [hunk 20]
```diff
@@ -0,0 +1,92 @@
...
+
+    @Test
+    public void isValidShouldReturnFalseWhenAboveMaxValue() {
+        assertThat(Port.isValid(Port.MAX_PORT_VALUE + 1))
+            .isFalse();
+    }
+
+    @Test
+    public void generateValidUnprivilegedPortShouldReturnAValidPort() {
+        assertThat(Port.generateValidUnprivilegedPort())
```
Comment (strokyl) [inline server/container/util-java8/src/test/java/org/apache/james/util/PortTest.java]: IMO we should test that more than once or not test it at all.

**Comment by chibenwa** [comment]
test this please

File: server/container/util-java8/src/test/java/org/apache/james/util/streams/ContainerNames.java [hunk 21]
```diff
@@ -25,4 +25,5 @@
     String ELASTICSEARCH = "elasticsearch:2.2.2";
     String NGINX = "nginx:1.7.1";
     String TIKA = "logicalspark/docker-tikaserver:1.15rc2";
+    String SPAMASSASSIN = "dinkel/spamassassin:3.4.0";
```
Comment (mbaechler) [inline server/container/util-java8/src/test/java/org/apache/james/util/streams/ContainerNames.java]: I discover that file : why it's called Names (they are not names) ?
And why in util.streams ?

File: server/mailet/mailets/pom.xml [hunk 22]
```diff
@@ -133,6 +133,19 @@
             <groupId>${project.groupId}</groupId>
             <artifactId>james-server-lifecycle-api</artifactId>
         </dependency>
+        <dependency>
+            <groupId>${project.groupId}</groupId>
+            <artifactId>james-server-protocols-library</artifactId>
```
Comment (mbaechler) [inline server/mailet/mailets/pom.xml]: it probably shows that tests are badly located.

File: server/protocols/protocols-smtp/src/test/java/org/apache/james/smtpserver/mock/util/MockSpamd.java [hunk 23]
```diff
@@ -82,16 +96,14 @@ public void run() {
...
-            in.close();
-            out.close();
-            spamd.close();
-            socket.close();
-
         } catch (IOException e) {
             // Should not happen
             e.printStackTrace();
+        } finally {
+            IOUtils.closeQuietly(in);
```
Comment (mbaechler) [inline server/protocols/protocols-smtp/src/test/java/org/apache/james/smtpserver/mock/util/MockSpamd.java]: autocloseable ?

File: server/protocols/webadmin/webadmin-core/src/main/java/org/apache/james/webadmin/PortProvider.java [hunk 24]
```diff
@@ -19,7 +19,7 @@
 
 package org.apache.james.webadmin;
 
-public interface Port {
+public interface PortProvider {
 
     int toInt();
```
Comment (mbaechler) [inline server/protocols/webadmin/webadmin-core/src/main/java/org/apache/james/webadmin/PortProvider.java]: this method is not relevant anymore. generatePort ? or providePort ? But then, we should keep a Port object, right ?

File: server/container/util-java8/src/test/java/org/apache/james/util/streams/ContainerNames.java (hunk 21 above)
Comment (chibenwa) [reply server/container/util-java8/src/test/java/org/apache/james/util/streams/ContainerNames.java]: It's next to SwarmContainer. Maybe we need a **docker** package in there. And maybe **Images** might be a better name for this.

Would you agree with such a change @mbaechler  ?

File: server/protocols/webadmin/webadmin-core/src/main/java/org/apache/james/webadmin/PortProvider.java (hunk 24 above)
Comment (chibenwa) [reply server/protocols/webadmin/webadmin-core/src/main/java/org/apache/james/webadmin/PortProvider.java]: Generate a port in test
Provide a port in prod.

Seriously, I don't intend to change the way webadmin works (as I don't concider it bad,just it's naming could have been improved).

File: server/mailet/mailets/pom.xml (hunk 22 above)
Comment (chibenwa) [reply server/mailet/mailets/pom.xml]: Moving the MockSpamd\* test helper stuff to james-testing stuff might solve this PB.

Would you agree with it @mbaechler ?

File: server/protocols/protocols-smtp/src/test/java/org/apache/james/smtpserver/mock/util/MockSpamd.java (hunk 23 above)
Comment (chibenwa) [reply server/protocols/protocols-smtp/src/test/java/org/apache/james/smtpserver/mock/util/MockSpamd.java]: I spend 1h debugging after doing that change. Is it essential?

File: server/container/util-java8/src/test/java/org/apache/james/util/streams/ContainerNames.java (hunk 21 above)
Comment (mbaechler) [reply server/container/util-java8/src/test/java/org/apache/james/util/streams/ContainerNames.java]: of course

File: server/protocols/protocols-smtp/src/test/java/org/apache/james/smtpserver/mock/util/MockSpamd.java (hunk 23 above)
Comment (mbaechler) [reply server/protocols/protocols-smtp/src/test/java/org/apache/james/smtpserver/mock/util/MockSpamd.java]: I don't understand what is so hard about autocloseable

Comment (chibenwa) [reply server/protocols/protocols-smtp/src/test/java/org/apache/james/smtpserver/mock/util/MockSpamd.java]: Me neither. As soon as I migrate this code, it breaks. Probably the order in the 'close' clauses.

**Comment by chibenwa** [comment]
Merged

**Pull request closed without merging**